Mathlib/Analysis/SpecialFunctions/Trigonometric/Chebyshev/RootsExtrema.lean

```diff
@@ -0,0 +1,235 @@
+    (T ℝ n).eval x ∈ Set.Icc (-1) 1 := by
+  rw [← cos_arccos (x := x) (by grind) (by grind)]
+  grind [T_real_cos, cos_mem_Icc]
+
+theorem T_real_abs_eval_le_one_of_abs_le_one (n : ℤ) {x : ℝ} (hx : |x| ≤ 1) :
+    |(T ℝ n).eval x| ≤ 1 := by
+  grind [T_real_eval_bounded_of_bounded]
+
+theorem T_real_one_le_eval_of_one_le (n : ℤ) {x : ℝ} (hx : 1 ≤ x) : 1 ≤ (T ℝ n).eval x := by
+  rw [← cosh_arcosh hx]
+  grind [T_real_cosh, one_le_cosh]
+
+theorem T_real_one_lt_eval_of_one_lt {n : ℤ} (hn : n ≠ 0) {x : ℝ} (hx : 1 < x) :
+    1 < (T ℝ n).eval x := by
+  have : arcosh x ≠ 0 := by
+    by_contra! h
+    grind [cosh_arcosh, cosh_zero]
+  rw [←cosh_arcosh (le_of_lt hx), T_real_cosh, one_lt_cosh, mul_ne_zero_iff]
+  exact ⟨by norm_cast, by assumption⟩
+
+theorem T_real_one_le_negOnePow_mul_eval_of_le_neg_one (n : ℤ) {x : ℝ} (hx : x ≤ -1) :
+    1 ≤ (-1) ^ n * (T ℝ n).eval x := by
+  rw [← neg_neg x, T_eval_neg]
+  convert T_real_one_le_eval_of_one_le n (le_neg_of_le_neg hx)
+  rw [Int.cast_negOnePow, ← mul_assoc, ← mul_zpow]
+  simp
+
+theorem T_real_one_lt_negOnePow_mul_eval_of_lt_neg_one {n : ℤ} (hn : n ≠ 0) {x : ℝ} (hx : x < -1) :
+    1 < (-1) ^ n * (T ℝ n).eval x := by
+  rw [← neg_neg x, T_eval_neg]
+  convert T_real_one_lt_eval_of_one_lt hn (lt_neg_of_lt_neg hx)
+  rw [Int.cast_negOnePow, ← mul_assoc, ← mul_zpow]
+  simp
+
+theorem T_real_one_le_abs_eval_of_one_le_abs (n : ℤ) {x : ℝ} (hx : 1 ≤ |x|) :
+    1 ≤ |(T ℝ n).eval x| := by
+  cases le_abs.mp hx with
+  | inl hx => exact T_real_one_le_eval_of_one_le n hx |> .inl |> le_abs.mpr
+  | inr hx =>
+    have := T_real_one_le_negOnePow_mul_eval_of_le_neg_one n (le_neg_of_le_neg hx)
+    have : 1 ≤ |(-1) ^ n * (T ℝ n).eval x| := by grind
+    convert this using 1
+    simp [abs_mul, abs_zpow]
+
+theorem T_real_one_lt_abs_eval_of_one_lt_abs {n : ℤ} (hn : n ≠ 0) {x : ℝ} (hx : 1 < |x|) :
+    1 < |(T ℝ n).eval x| := by
+  cases lt_abs.mp hx with
+  | inl hx => exact T_real_one_lt_eval_of_one_lt hn hx |> .inl |> lt_abs.mpr
+  | inr hx =>
+    have := T_real_one_lt_negOnePow_mul_eval_of_lt_neg_one hn (lt_neg_of_lt_neg hx)
+    have : 1 < |(-1) ^ n * (T ℝ n).eval x| := by grind
+    convert this using 1
+    simp [abs_mul, abs_zpow]
+
+theorem T_real_abs_eval_le_one_iff_abs_le_one {n : ℤ} (hn : n ≠ 0) (x : ℝ) :
+    |x| ≤ 1 ↔ |(T ℝ n).eval x| ≤ 1 := by
+  constructor
+  · intro hx; exact T_real_abs_eval_le_one_of_abs_le_one n hx
+  · intro hx; contrapose! hx; exact T_real_one_lt_abs_eval_of_one_lt_abs hn hx
+
+theorem T_real_eval_eq_cos_of_abs_le_one {n : ℤ} (hn : n ≠ 0) {x : ℝ} (hx : |(T ℝ n).eval x| ≤ 1) :
+    (T ℝ n).eval x = cos (n * arccos x) := by
+  have := (T_real_abs_eval_le_one_iff_abs_le_one hn x).mpr hx
+  rw [← T_real_cos, cos_arccos (by grind) (by grind)]
+
+
+/-- `T_real_roots n` is the set of roots of `T ℝ n`. -/
+noncomputable def T_real_roots (n : ℕ) : Finset ℝ :=
+  (Finset.range n).image (fun (k : ℕ) => cos ((2 * k + 1) * π / (2 * n)))
```

Comment:
Why do you want this to be a `def`? I could see it *maybe* being an `abbrev`, but I'm not sure that's worth it. I would have expected the declaration with this name (or rather `roots_T_real`) to be the lemma:
```lean
example : (T ℝ n).roots = (Finset.range n).image (fun (k : ℕ) => cos ((2 * k + 1) * π / (2 * n)))
```
Also, if you really do want this to be a `def`, it needs to be `lowerCamelCase` at the very least in order to adhere to our naming conventions.